Assets/BossRoom/Scripts/Server/Net/ServerSessionManager.cs

```diff
@@ -0,0 +1,272 @@
+                    //to remember client information on a per-guid basis after the connection has been lost.
+                    m_ClientData.Remove(guid);
+                }
+            }
+
+            if (clientId == m_Portal.NetManager.LocalClientId)
+            {
+                //the ServerGameNetPortal may be initialized again, which will cause its OnNetworkSpawn to be called again.
+                //Consequently we need to unregister anything we registered, when the NetworkManager is shutting down.
+                m_Portal.UserDisconnectRequested -= OnUserDisconnectRequest;
+                m_Portal.NetManager.OnClientDisconnectCallback -= OnClientDisconnect;
+            }
+        }
+
+
+        /// <summary>
+        /// Handles the flow when a user has requested a disconnect via UI (which can be invoked on the Host, and thus must be
+        /// handled in server code).
+        /// </summary>
+        private void OnUserDisconnectRequest()
+        {
+            Clear();
+        }
+
+        private void Clear()
+        {
+            //resets all our runtime state.
+            m_ClientData.Clear();
+            m_ClientIDToGuid.Clear();
+        }
+
+
+        /// <summary>
+        /// This logic plugs into the "ConnectionApprovalCallback" exposed by MLAPI.NetworkManager, and is run every time a client connects to us.
+        /// See GNH_Client.StartClient for the complementary logic that runs when the client starts its connection.
+        /// </summary>
+        /// <remarks>
+        /// Since our game doesn't have to interact with some third party authentication service to validate the identity of the new connection, our ApprovalCheck
+        /// method is simple, and runs synchronously, invoking "callback" to signal approval at the end of the method. MLAPI currently doesn't support the ability
+        /// to send back more than a "true/false", which means we have to work a little harder to provide a useful error return to the client. To do that, we invoke a
+        /// client RPC in the same channel that MLAPI uses for its connection callback. Since that channel ("MLAPI_INTERNAL") is both reliable and sequenced, we can be
+        /// confident that our login result message will execute before any disconnect message.
+        /// </remarks>
+        /// <param name="connectionData">binary data passed into StartClient. In our case this is the client's GUID, which is a unique identifier for their install of the game that persists across app restarts. </param>
+        /// <param name="clientId">This is the clientId that MLAPI assigned us on login. It does not persist across multiple logins from the same client. </param>
+        /// <param name="callback">The delegate we must invoke to signal that the connection was approved or not. </param>
+        private void ApprovalCheck(byte[] connectionData, ulong clientId, MLAPI.NetworkManager.ConnectionApprovedDelegate callback)
```

Comment:
wait, there's two ApprovalChecks?
mmm looking further, looks like callback isn't used. So basically, we're using this one here to insert ourselves during the connection approval process? Isn't there another callback we could use for this, like OnClientConnectedCallback?